components/core/src/clp/NetworkReader.cpp

```diff
@@ -0,0 +1,354 @@
+        auto const data_view{data.subspan(0, std::min(curr_downloader_buf.size(), data.size()))};
+        std::copy(data_view.begin(), data_view.end(), curr_downloader_buf.begin());
+        data = data.subspan(data_view.size());
+        curr_downloader_buf = curr_downloader_buf.subspan(data_view.size());
+        if (curr_downloader_buf.empty()) {
+            enqueue_filled_buffer();
+        }
+    }
+    return num_bytes_to_write;
+}
+
+auto NetworkReader::DownloaderThread::thread_method() -> void {
+    try {
+        CurlDownloadHandler curl_handler{
+                curl_progress_callback,
+                curl_write_callback,
+                static_cast<void*>(&m_reader),
+                m_reader.m_src_url,
+                m_offset,
+                m_disable_caching,
+                m_reader.m_connection_timeout,
+                m_reader.m_overall_timeout
+        };
+        auto const retcode{curl_handler.perform()};
+        m_reader.m_atomic_curl_return_code.store(retcode);
+        // Enqueue the last filled buffer, if any
+        m_reader.enqueue_filled_buffer();
+        m_reader.set_state_code((CURLE_OK == retcode) ? State::Finished : State::Failed);
+    } catch (CurlOperationFailed const& ex) {
+        m_reader.m_atomic_curl_return_code.store(ex.get_curl_err());
+        m_reader.set_state_code(State::Failed);
+    }
+
+    std::unique_lock<std::mutex> const buffer_resource_lock{m_reader.m_buffer_resource_mutex};
+    m_reader.m_cv_reader.notify_all();
+}
+
+auto NetworkReader::abort_data_download() -> void {
+    m_download_aborted.store(true);
+
+    std::unique_lock<std::mutex> const buffer_resource_lock{m_buffer_resource_mutex};
+    m_cv_downloader.notify_all();
+}
+
+auto NetworkReader::acquire_empty_buffer() -> void {
+    if (m_curr_downloader_buf.has_value()) {
+        return;
+    }
+    std::unique_lock<std::mutex> buffer_resource_lock{m_buffer_resource_mutex};
+    while (m_filled_buffer_queue.size() == m_buffer_pool_size) {
+        m_cv_downloader.wait(buffer_resource_lock);
+        if (is_download_aborted() || is_download_timedout()) {
+            return;
+        }
+    }
+    m_curr_downloader_buf.emplace(m_buffer_pool.at(m_curr_downloader_buf_idx).get(), m_buffer_size);
+}
+
+auto NetworkReader::enqueue_filled_buffer() -> void {
+    if (false == m_curr_downloader_buf.has_value()) {
+        return;
+    }
+    std::unique_lock<std::mutex> const buffer_resource_lock{m_buffer_resource_mutex};
+    m_filled_buffer_queue.emplace(
+            m_buffer_pool.at(m_curr_downloader_buf_idx).get(),
+            m_buffer_size - m_curr_downloader_buf.value().size()
+    );
+
+    m_curr_downloader_buf.reset();
+    ++m_curr_downloader_buf_idx;
+    if (m_curr_downloader_buf_idx == m_buffer_pool_size) {
+        m_curr_downloader_buf_idx = 0;
+    }
+
+    m_cv_reader.notify_all();
+}
+
+auto NetworkReader::get_filled_buffer() -> void {
```

Comment:
We didn't dequeue the buffer. Dequeuing happens when we release the empty buffer. Writing this way, we can use the queue size to indicate whether there're any free buffers for the downloader. If the queue size equals the pool size, then there will be no empty buffer available for write. Otherwise, we need extra member variables to keep track of number of non-empty buffers.

Reply:
True, true. My mistake. You'll need to update the diagram then.